include/fiction/technology/charge_distribution_surface.hpp

```diff
@@ -828,10 +1496,455 @@ class charge_distribution_surface<Lyt, false> : public Lyt
+     * @param previous_gray_code gray code in decimal representing the old charge distribution.
+     * @param dependent_cell if set to VARIABLE, the dependent cell's charge state is changed based on the local
+     * potential at its position.
+     * @param energy_calc_mode if set to UPDATE_ENERGY, the system energy is calculated for the new charge distribution.
+     * @param history_mode if set to NEGLECT, the values of the previous charge distribution are not used to speed up
+     * the calculation.
+     */
+    void assign_charge_index_by_gray_code(
+        const uint64_t current_gray_code, const uint64_t previous_gray_code,
+        const dependent_cell_mode&         dependent_cell   = dependent_cell_mode::FIXED,
+        const energy_calculation&          energy_calc_mode = energy_calculation::UPDATE_ENERGY,
+        const charge_distribution_history& history_mode     = charge_distribution_history::NEGLECT) noexcept
+    {
+        if (current_gray_code <= strg->max_charge_index)
+        {
+            this->assign_charge_index_by_two_gray_codes(current_gray_code, previous_gray_code);
+            this->update_after_charge_change(dependent_cell, energy_calc_mode, history_mode);
+        }
+    }
+    /**
+     * Resets the charge index of the sublayout (cells of the layout that can also be positively charged).
+     */
+    void reset_charge_index_sub_layout() noexcept
+    {
+        strg->charge_index_sublayout.first = 0;
+        this->index_to_charge_distribution(exhaustive_algorithm::QUICKEXACT);
+        this->update_after_charge_change(dependent_cell_mode::VARIABLE, energy_calculation::KEEP_OLD_ENERGY_VALUE,
+                                         charge_distribution_history::CONSIDER);
+    }
+    /**
+     * Returns the maximum index of the sublayout (cells that can be positively charged).
+     *
+     * @returns The maximal possible charge distribution index of the sublayout.
+     */
+    [[nodiscard]] uint64_t get_max_charge_index_sub_layout() const noexcept
+    {
+        return strg->max_charge_index_sulayout;
+    }
+    /**
+     * Assign a given charge index to the charge distribution layout. This function should be used if new and old charge
+     * index are given as gray code to provide high performance.
+     *
+     * @param gray_code charge index (as gray code in decimal) of the new charge distribution.
+     * @param gray_code_old charge index (as gray code in decimal) of the old charge distribution.
+     */
+    void assign_charge_index_by_two_gray_codes(const uint64_t gray_code, const uint64_t gray_code_old) noexcept
+    {
+        strg->charge_index.first = gray_code;
+        this->charge_index_gray_code_to_charge_distribution(gray_code, gray_code_old);
+    }
+    /**
+     * This function returns all SiDBs of the layout.
+     *
+     * @return Vector with all cells.
+     */
+    std::vector<typename Lyt::cell> get_sidb_order() noexcept
+    {
+        return strg->sidb_order;
+    }
```

Comment:
Why is this function required? There is `get_all_sidb_cells` already, which is not used, btw. 

Generally, I would strongly avoid having getters that return complex objects as mentioned above.